File: Example/PayTPV Objc Example/AddCardController.m
//
//  AddCardController.m
//  PAYTPV Objc Example
//
//  Created by Mihail Cristian Dumitru on 8/21/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

#import "AddCardController.h"
#import "AlertBuilder.h"
#import "Models.h"
#import "ModelStore.h"

@interface AddCardController () <UITextFieldDelegate>

#pragma mark - IBOutlets

@property (weak, nonatomic) IBOutlet UITextField *panTextField;
@property (weak, nonatomic) IBOutlet UITextField *monthTextField;
@property (weak, nonatomic) IBOutlet UITextField *yearTextField;
@property (weak, nonatomic) IBOutlet UITextField *cvvTextField;

@end

@implementation AddCardController

#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setup];
}

#pragma mark - Setup

- (void)setup {
    self.title = @"Add a card";

    [self setupTextFields];
}

- (void)setupTextFields {
    self.panTextField.delegate = self;
    self.monthTextField.delegate = self;
    self.yearTextField.delegate = self;
    self.cvvTextField.delegate = self;
}

#pragma mark - Helpers

- (NSString *)expiryDate {
    return [NSString stringWithFormat:@"%@%@", self.monthTextField.text, self.yearTextField.text];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:true];
}

- (BOOL)string:(NSString *)string
  matchesRegex:(NSString *)regexString {
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:regexString
                                                                           options:NSRegularExpressionCaseInsensitive
                                                                             error:nil];
    return [regex numberOfMatchesInString:string options:0 range:NSMakeRange(0, string.length)];
}

- (BOOL)validPan:(NSString *)pan {
    return [self string:pan matchesRegex:@"^[0-9]{0,19}$"];
}

- (BOOL)validMonth:(NSString *)month {
    return [self string:month matchesRegex:@"(^0[1-9]?$)|(^1[012]?$)"];
}

- (BOOL)validYear:(NSString *)year {
    return [self string:year matchesRegex:@"(1[6-9]?$)|(^[2-9][0-9]?$)"];
}

- (BOOL)validCvv:(NSString *)cvv {
    return [self string:cvv matchesRegex:@"^[0-9]{0,4}$"];
}

#pragma mark - IBActions

- (IBAction)onCancelButtonPressed:(UIButton *)sender {
    [self dismissViewControllerAnimated:true completion:nil];
}

- (IBAction)onAddCardButtonPressed:(UIButton *)sender {
    // add a loading indicator
    AlertBuilder *loader = [AlertBuilder new];
    [loader setTitle:@"Adding card"];
    [loader addLoadingIndicator];
    [loader presentInController:self animated:true completion:nil];

    // make the request to PAYTPV
    PTPVCard *card = [[PTPVCard alloc] initWithPan:self.panTextField.text
                                    expiryDate:[self expiryDate]
                                           cvv:self.cvvTextField.text];
    [[PTPVAPIClient sharedClient] addUser:card completion:^(PTPVUser * _Nullable user, NSError * _Nullable error) {
        [loader dismissAnimated:true completion:^{
            if (user == nil) {
                if (error != nil) {
                    AlertBuilder *alert = [AlertBuilder new];
                    [alert setMessage:error.localizedDescription];
                    [alert addActionWithTitle:@"OK"
                                        style:UIAlertActionStyleDefault
                                      handler:nil];
                    [alert presentInController:self animated:true completion:nil];
                    return;
                }
                AlertBuilder *alert = [AlertBuilder new];
                [alert setMessage:@"There was an unexpected error. Please try again later."];
                [alert addActionWithTitle:@"OK"
                                    style:UIAlertActionStyleDefault
                                  handler:nil];
                [alert presentInController:self animated:true completion:nil];
                return;
            }

            // send the user token to the backend
            Card *card = [[Card alloc] initWithIdUser:user.idUser tokenUser:user.tokenUser name:self.panTextField.text];
            [ModelStore addCard:card completion:^{
                [self dismissViewControllerAnimated:true completion:^{
                    if (self.delegate) {
                        [self.delegate addCardController:self didFinishAddingCard:card];
                    }
                }];
            }];
        }];
    }];
}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    NSUInteger newLength = [textField.text length] + [string length] - range.length;
    if(newLength == 0) return YES;

    NSString *newString = [textField.text stringByReplacingCharactersInRange:range withString:string];

    if (textField == self.panTextField) {
        return [self validPan:newString];
    } else if (textField == self.monthTextField) {
        return [self validMonth:newString];
    } else if (textField == self.yearTextField) {
        return [self validYear:newString];
    } else if (textField == self.cvvTextField) {
        return [self validCvv:newString];
    }

    return true;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == self.panTextField) {
        [self.monthTextField becomeFirstResponder];
    } else if (textField == self.monthTextField) {
        [self.yearTextField becomeFirstResponder];
    } else if (textField == self.yearTextField) {
        [self.cvvTextField becomeFirstResponder];
    }

    return true;
}

@end

File: Example/PayTPV Swift Example/AddCardController.swift
//
//  AddCardController.swift
//  PAYTPV Swift Example
//
//  Created by Mihail Cristian Dumitru on 8/18/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

import UIKit
import PAYTPV

@objc protocol AddCardDelegate {
    func addCardController(_ controller: AddCardController, didFinishAddingCard card: Card)
}

class AddCardController: UIViewController {

    // MARK: - IBOutlets

    @IBOutlet weak var panTextField: UITextField!
    @IBOutlet weak var monthTextField: UITextField!
    @IBOutlet weak var yearTextField: UITextField!
    @IBOutlet weak var cvvTextField: UITextField!

    // MARK: - Properties

    weak var delegate: AddCardDelegate?

    var pan: String {
        return self.panTextField.text ?? ""
    }

    var expiryDate: String {
        return "\(self.monthTextField.text ?? "")\(self.yearTextField.text ?? "")"
    }

    var cvv: String {
        return self.cvvTextField.text ?? ""
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        self.setup()
    }

    // MARK: - Setup

    func setup() {
        self.title = "Add a card"

        self.setupTextFields()
    }

    func setupTextFields() {
        self.panTextField.delegate = self
        self.monthTextField.delegate = self
        self.yearTextField.delegate = self
        self.cvvTextField.delegate = self
    }

    // MARK: - Helpers

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }

    func string(string: String, matchesRegex regexString: String) -> Bool {
        do {
            let regex = try NSRegularExpression(pattern: regexString, options: .caseInsensitive)

            return regex.numberOfMatches(in: string, options: [], range: NSRange(location: 0, length: string.characters.count)) > 0
        } catch (_) {

        }

        return true
    }

    func valid(pan: String) -> Bool {
        return self.string(string: pan, matchesRegex: "^[0-9]{0,19}$")
    }

    func valid(month: String) -> Bool {
        return self.string(string: month, matchesRegex: "(^0[1-9]?$)|(^1[012]?$)")
    }

    func valid(year: String) -> Bool {
        return self.string(string: year, matchesRegex: "(1[6-9]?$)|(^[2-9][0-9]?$)")
    }

    func valid(cvv: String) -> Bool {
        return self.string(string: cvv, matchesRegex: "^[0-9]{0,4}$")
    }

    // MARK: - IBActions

    @IBAction func onAddCardButtonPressed(_ sender: UIButton) {
        // add a loading indicator
        let loader = AlertBuilder()
            .set(title: "Adding card")
            .addLoadingIndicator()
            .present(sender: self)

        // make the request to PAYTPV
        let card = PTPVCard(pan: self.pan, expiryDate: self.expiryDate, cvv: self.cvv)

        PTPVAPIClient.shared().addUser(card) { (user, error) in
            loader.dismiss(animated: true, completion: { 
                guard let user = user else {
                    if let error = error {
                        let _ = AlertBuilder()
                            .set(message: error.localizedDescription)
                            .add(actionWithTitle: "OK")
                            .present(sender: self)
                        return;
                    }
                    let _ = AlertBuilder()
                        .set(message: "There was an unexpected error. Please try again later.")
                        .add(actionWithTitle: "OK")
                        .present(sender: self)
                    return;
                }

                // send the user token to the backend
                let card = Card(id: user.idUser ?? "", token: user.tokenUser ?? "", name: self.pan)
                ModelStore.add(card: card, completion: { (_) in
                    self.dismiss(animated: true, completion: {
                        self.delegate?.addCardController(self, didFinishAddingCard: card)
                    })
                })
            })
        }
    }

    @IBAction func onCancelButtonPressed(_ sender: UIButton) {
        self.dismiss(animated: true, completion: nil)
    }

}

// MARK: - UITextFieldDelegate

extension AddCardController: UITextFieldDelegate {

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {

        guard let text = textField.text as NSString? else {
            return true
        }

        let newLength = text.length + string.characters.count - range.length
        if newLength == 0 {
            return true
        }

        let newString = text.replacingCharacters(in: range, with: string)

        switch textField {
        case self.panTextField:
            return self.valid(pan: newString)

        case self.monthTextField:
            return self.valid(month: newString)

        case self.yearTextField:
            return self.valid(year: newString)

        case self.cvvTextField:
            return self.valid(cvv: newString)

        default: break
        }

        return true
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        switch textField {
        case self.panTextField:
            self.monthTextField.becomeFirstResponder()

        case self.monthTextField:
            self.yearTextField.becomeFirstResponder()

        case self.yearTextField:
            self.cvvTextField.becomeFirstResponder()
            
        default: break
        }
        
        return true
    }
    
}
